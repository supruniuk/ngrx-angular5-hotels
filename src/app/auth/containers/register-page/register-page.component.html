<mat-drawer-container>
  <div fxLayout="row" fxLayoutAlign="center center" class="register-container">
    <hot-register-form
      (onSubmit)="onSubmit($event)"
      (onCancel)="onCancel()"
      [pending]="pending$ | async"
      [errorMessage]="error$ | async">
    </hot-register-form>
  </div>
</mat-drawer-container>