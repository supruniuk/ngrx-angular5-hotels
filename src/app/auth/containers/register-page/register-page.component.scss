:host {
  height:100%;
  display:flex;
  flex-direction: column;
}

.register-container {
  height:100%;
}

mat-drawer-container {
  height: 100%;
}