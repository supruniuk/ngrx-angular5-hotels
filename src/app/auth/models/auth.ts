import { User } from './user';


export interface Authenticate {
  email: string;
  password: string;
}

export interface SignUp extends Authenticate {
  authType: string;
  name: string;
}

export interface AuthResponse {
  user: User;
  token: string;
}

export const CUSTOMER_TYPE = 'customer';
export const HOTEL_TYPE = 'hotel';

export class Auth {
  constructor(
    public token: string,
    public type: string
  ) { }
}

// export class CustomerAuth extends Auth {
//   readonly type = CUSTOMER_TYPE;
// }

// export class HotelAuth extends Auth {
//   readonly type = HOTEL_TYPE;
// }
