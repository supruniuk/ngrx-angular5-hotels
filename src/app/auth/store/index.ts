export * from './actions/auth';
export * from './reducers';
export * from './effects/auth.effects';
export * from './selectors';
