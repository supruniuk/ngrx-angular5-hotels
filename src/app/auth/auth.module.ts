import { NgModule, ModuleWithProviders } from '@angular/core';
import { CommonModule } from '@angular/common';
import { RouterModule } from '@angular/router';
import { HTTP_INTERCEPTORS } from '@angular/common/http';
import { ReactiveFormsModule } from '@angular/forms';
import { FlexLayoutModule } from '@angular/flex-layout';
import { StoreModule } from '@ngrx/store';
import { EffectsModule } from '@ngrx/effects';
import { Routes } from '@angular/router';

import { MaterialModule } from '../material';
import { LoginPageComponent } from './containers/login-page/login-page.component';
import { RegisterPageComponent } from './containers/register-page/register-page.component';
import { LoginFormComponent } from './components/login-form/login-form.component';
import { RegisterFormComponent } from './components/register-form/register-form.component';

import { AuthInterceptor, AuthService, AuthGuard, RoleGuard } from './services';
import { AuthEffects, reducers, metaReducers } from './store';


export const COMPONENTS = [
  LoginPageComponent,
  LoginFormComponent,
  RegisterPageComponent,
  RegisterFormComponent,
];

export const ROUTES: Routes = [
  { path: 'login', component: LoginPageComponent },
  { path: 'register', component: RegisterPageComponent }
];

@NgModule({
  imports: [
    CommonModule,
    ReactiveFormsModule,
    MaterialModule,
    FlexLayoutModule
  ],
  declarations: COMPONENTS,
  exports: COMPONENTS,
})
export class AuthModule {
  static forRoot(): ModuleWithProviders {
    return {
      ngModule: RootAuthModule,
      providers: [
        AuthService,
        AuthGuard,
        RoleGuard,
        {
          provide: HTTP_INTERCEPTORS,
          useClass: AuthInterceptor,
          multi: true
        },
      ],
    };
  }
}

@NgModule({
  imports: [
    AuthModule,
    RouterModule.forChild(ROUTES),
    StoreModule.forFeature('auth', reducers, { metaReducers }),
    EffectsModule.forFeature([AuthEffects]),
  ],
})
export class RootAuthModule { }
