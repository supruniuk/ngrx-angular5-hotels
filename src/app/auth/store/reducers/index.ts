import { localStorageSync } from 'ngrx-store-localstorage';
import {
  Action,
  ActionReducer,
  createFeatureSelector,
  createSelector,
  MetaReducer
} from '@ngrx/store';
import { RootState } from '../../../core/store/reducers';
import * as fromStatus from './auth-status.reducers';
import * as fromLoginPage from './login-page.reducers';
import * as fromRegisterPage from './register-page.reducers';


export interface AuthState {
  status: fromStatus.StatusState;
  loginPage: fromLoginPage.LoginState;
  registerPage: fromRegisterPage.RegisterState;
}

export interface RootAuthState extends RootState {
  auth: AuthState;
}

export const reducers = {
  status: fromStatus.reducer,
  loginPage: fromLoginPage.reducer,
  registerPage: fromRegisterPage.reducer,
};

// Setting up ngrx-store-localstorage reducer
// to keep authentication info in the LocalStorage
export function localStorageSyncReducer(reducer: ActionReducer<any>): ActionReducer<any> {
  return localStorageSync({ keys: ['status'], rehydrate: true })(reducer);
}

export const metaReducers: Array<MetaReducer<any, any>> = [localStorageSyncReducer];

export const getAuthState = createFeatureSelector<AuthState>('auth');
