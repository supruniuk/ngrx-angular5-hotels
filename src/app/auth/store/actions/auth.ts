import { Action } from '@ngrx/store';
import { User } from '../../models/user';
import { Authenticate, SignUp, Auth } from '../../models/auth';

export enum AuthActionTypes {
  login = '[Auth] Login',
  logout = '[Auth] Logout',
  loginSuccess = '[Auth] Login Success',
  loginFail = '[Auth] Login Failure',
  loginRedirect = '[Auth] Login Redirect',
  customerRedirect = '[Auth] Customer Redirect',
  hotelRedirect = '[Auth] Hotel Redirect',
  register = '[Auth] Register',
  registerSuccess = '[Auth] Register Success',
  registerFail = '[Auth] Register Failure',
  registerRedirect = '[Auth] Register Redirect',
}

export class Login implements Action {
  readonly type = AuthActionTypes.login;

  constructor(public payload: Authenticate) {}
}

export class LoginSuccess implements Action {
  readonly type = AuthActionTypes.loginSuccess;

  constructor(public payload: { user: User, auth: Auth }) {}
}

export class LoginFail implements Action {
  readonly type = AuthActionTypes.loginFail;

  constructor(public payload: any) {}
}

export class LoginRedirect implements Action {
  readonly type = AuthActionTypes.loginRedirect;
}

export class CustomerRedirect implements Action {
  readonly type = AuthActionTypes.customerRedirect;
}

export class HotelRedirect implements Action {
  readonly type = AuthActionTypes.hotelRedirect;
}

export class Register implements Action {
  readonly type = AuthActionTypes.register;

  constructor(public payload: SignUp) {}
}

export class RegisterSuccess implements Action {
  readonly type = AuthActionTypes.registerSuccess;

  constructor(public payload: { user: User, auth: Auth }) {}
}

export class RegisterFail implements Action {
  readonly type = AuthActionTypes.registerFail;

  constructor(public payload: any) {}
}

export class RegisterRedirect implements Action {
  readonly type = AuthActionTypes.registerRedirect;

  constructor(public payload: string) {}
}

export class Logout implements Action {
  readonly type = AuthActionTypes.logout;
}

export type AuthActions = Login
  | LoginSuccess
  | LoginFail
  | LoginRedirect
  | CustomerRedirect
  | HotelRedirect
  | Register
  | RegisterSuccess
  | RegisterFail
  | RegisterRedirect
  | Logout;
