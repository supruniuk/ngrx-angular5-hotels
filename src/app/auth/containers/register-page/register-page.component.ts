import { Component, OnInit } from '@angular/core';
import { Store } from '@ngrx/store';
import { SignUp } from '../../models/auth';
import { AuthState } from '../../store/reducers';
import { getRegisterPagePending, getRegisterPageError } from '../../store/selectors';
import { Register, LoginRedirect } from '../../store/actions/auth';


@Component({
  selector: 'hot-register-page',
  templateUrl: './register-page.component.html',
  styleUrls: ['./register-page.component.scss'],
})
export class RegisterPageComponent implements OnInit {
  pending$ = this.store.select(getRegisterPagePending);
  error$ = this.store.select(getRegisterPageError);

  constructor(private store: Store<AuthState>) { }

  ngOnInit() { }

  onSubmit($event: SignUp) {
    this.store.dispatch(new Register($event));
  }

  onCancel() {
    this.store.dispatch(new LoginRedirect());
  }
}
