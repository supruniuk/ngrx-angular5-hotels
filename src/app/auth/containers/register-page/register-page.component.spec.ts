import { TestBed, ComponentFixture } from '@angular/core/testing';
import { MatInputModule, MatCardModule } from '@angular/material';
import { ReactiveFormsModule } from '@angular/forms';
import { NoopAnimationsModule } from '@angular/platform-browser/animations';
import { StoreModule, Store, combineReducers } from '@ngrx/store';
import { RegisterPageComponent } from './register-page.component';
import { RegisterFormComponent } from '../../components/register-form/register-form.component';
import * as Auth from '../../actions/auth';
import * as fromAuth from '../../reducers';

describe('Register Page', () => {
  let fixture: ComponentFixture<RegisterPageComponent>;
  let store: Store<fromAuth.State>;
  let instance: RegisterPageComponent;

  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [
        NoopAnimationsModule,
        StoreModule.forRoot({
          auth: combineReducers(fromAuth.reducers),
        }),
        MatInputModule,
        MatCardModule,
        ReactiveFormsModule,
      ],
      declarations: [RegisterPageComponent, RegisterFormComponent],
    });

    fixture = TestBed.createComponent(RegisterPageComponent);
    instance = fixture.componentInstance;
    store = TestBed.get(Store);

    spyOn(store, 'dispatch').and.callThrough();
  });

  /**
   * Container components are used as integration points for connecting
   * the store to presentational components and dispatching
   * actions to the store.
   *
   * Container methods that dispatch events are like a component's output observables.
   * Container properties that select state from store are like a component's input properties.
   * If pure components are functions of their inputs, containers are functions of state
   *
   * Traditionally you would query the components rendered template
   * to validate its state. Since the components are analagous to
   * pure functions, we take snapshots of these components for a given state
   * to validate the rendered output and verify the component's output
   * against changes in state.
   */
  it('should compile', () => {
    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

  it('should dispatch a REGISTER action on submit', () => {
    const $event: any = {};
    const action = new Auth.Register($event);

    instance.onSubmit($event);

    expect(store.dispatch).toHaveBeenCalledWith(action);
  });

  it('should dispatch a LOGIN_REDIRECT action on cancel', () => {
    const action = new Auth.LoginRedirect();

    instance.onCancel();

    expect(store.dispatch).toHaveBeenCalledWith(action);
  });

});
