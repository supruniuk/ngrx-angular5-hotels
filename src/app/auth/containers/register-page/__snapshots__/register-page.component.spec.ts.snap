// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register Page should compile 1`] = `
<hot-register-page
  error$={[Function Store]}
  pending$={[Function Store]}
  store={[Function Store]}
>
  <hot-register-form
    ng-reflect-pending="false"
  >
    <mat-card
      class="mat-card"
    >
      
    
      <mat-card-title
        class="mat-card-title"
      >
        Register
      </mat-card-title>
      
    
      <mat-card-content
        class="mat-card-content"
      >
        
      
        <form
          class="ng-untouched ng-pristine ng-invalid"
          ng-reflect-form="[object Object]"
          novalidate=""
        >
          
  
        
          <p>
            
          
            <mat-form-field
              class="mat-input-container mat-form-field ng-tns-c1-0 ng-untouched ng-pristine ng-invalid"
            >
              <div
                class="mat-input-wrapper mat-form-field-wrapper"
              >
                <div
                  class="mat-input-flex mat-form-field-flex"
                >
                  
                  <div
                    class="mat-input-infix mat-form-field-infix"
                  >
                    
            
                    <input
                      aria-invalid="false"
                      class="mat-input-element mat-form-field-autofill-control ng-untouched ng-pristine ng-invalid"
                      formcontrolname="name"
                      id="mat-input-0"
                      matInput=""
                      ng-reflect-name="name"
                      ng-reflect-placeholder="Name"
                      ng-reflect-required=""
                      ng-reflect-type="text"
                      placeholder="Name"
                      required=""
                      type="text"
                    />
                    
          
                    <span
                      class="mat-input-placeholder-wrapper mat-form-field-placeholder-wrapper"
                    >
                      
                      <label
                        aria-owns="mat-input-0"
                        class="mat-input-placeholder mat-form-field-placeholder ng-tns-c1-0 mat-empty mat-form-field-empty mat-float mat-form-field-float"
                        for="mat-input-0"
                      >
                        Name 
                        
                        <span
                          aria-hidden="true"
                          class="mat-placeholder-required mat-form-field-required-marker ng-tns-c1-0"
                        >
                          *
                        </span>
                      </label>
                    </span>
                  </div>
                  
                </div>
                <div
                  class="mat-input-underline mat-form-field-underline"
                >
                  <span
                    class="mat-input-ripple mat-form-field-ripple"
                  />
                </div>
                <div
                  class="mat-input-subscript-wrapper mat-form-field-subscript-wrapper"
                  ng-reflect-ng-switch="hint"
                >
                  
                  
                  <div
                    class="mat-input-hint-wrapper mat-form-field-hint-wrapper ng-tns-c1-0 ng-trigger ng-trigger-transitionMessages"
                    style="opacity: 1;"
                  >
                    
                    <div
                      class="mat-input-hint-spacer mat-form-field-hint-spacer"
                    />
                  </div>
                </div>
              </div>
            </mat-form-field>
            
        
          </p>
          
  
        
          <p>
            
          
            <mat-form-field
              class="mat-input-container mat-form-field ng-tns-c1-1 ng-untouched ng-pristine ng-invalid"
            >
              <div
                class="mat-input-wrapper mat-form-field-wrapper"
              >
                <div
                  class="mat-input-flex mat-form-field-flex"
                >
                  
                  <div
                    class="mat-input-infix mat-form-field-infix"
                  >
                    
            
                    <input
                      aria-invalid="false"
                      class="mat-input-element mat-form-field-autofill-control ng-untouched ng-pristine ng-invalid"
                      formcontrolname="email"
                      id="mat-input-1"
                      matInput=""
                      ng-reflect-name="email"
                      ng-reflect-placeholder="Email"
                      ng-reflect-required=""
                      ng-reflect-type="text"
                      placeholder="Email"
                      required=""
                      type="text"
                    />
                    
          
                    <span
                      class="mat-input-placeholder-wrapper mat-form-field-placeholder-wrapper"
                    >
                      
                      <label
                        aria-owns="mat-input-1"
                        class="mat-input-placeholder mat-form-field-placeholder ng-tns-c1-1 mat-empty mat-form-field-empty mat-float mat-form-field-float"
                        for="mat-input-1"
                      >
                        Email 
                        
                        <span
                          aria-hidden="true"
                          class="mat-placeholder-required mat-form-field-required-marker ng-tns-c1-1"
                        >
                          *
                        </span>
                      </label>
                    </span>
                  </div>
                  
                </div>
                <div
                  class="mat-input-underline mat-form-field-underline"
                >
                  <span
                    class="mat-input-ripple mat-form-field-ripple"
                  />
                </div>
                <div
                  class="mat-input-subscript-wrapper mat-form-field-subscript-wrapper"
                  ng-reflect-ng-switch="hint"
                >
                  
                  
                  <div
                    class="mat-input-hint-wrapper mat-form-field-hint-wrapper ng-tns-c1-1 ng-trigger ng-trigger-transitionMessages"
                    style="opacity: 1;"
                  >
                    
                    <div
                      class="mat-input-hint-spacer mat-form-field-hint-spacer"
                    />
                  </div>
                </div>
              </div>
            </mat-form-field>
            
        
          </p>
          
  
        
          <p>
            
          
            <mat-form-field
              class="mat-input-container mat-form-field ng-tns-c1-2 ng-untouched ng-pristine ng-invalid"
            >
              <div
                class="mat-input-wrapper mat-form-field-wrapper"
              >
                <div
                  class="mat-input-flex mat-form-field-flex"
                >
                  
                  <div
                    class="mat-input-infix mat-form-field-infix"
                  >
                    
            
                    <input
                      aria-invalid="false"
                      class="mat-input-element mat-form-field-autofill-control ng-untouched ng-pristine ng-invalid"
                      formcontrolname="password"
                      id="mat-input-2"
                      matInput=""
                      ng-reflect-name="password"
                      ng-reflect-placeholder="Password"
                      ng-reflect-required=""
                      ng-reflect-type="password"
                      placeholder="Password"
                      required=""
                      type="password"
                    />
                    
          
                    <span
                      class="mat-input-placeholder-wrapper mat-form-field-placeholder-wrapper"
                    >
                      
                      <label
                        aria-owns="mat-input-2"
                        class="mat-input-placeholder mat-form-field-placeholder ng-tns-c1-2 mat-empty mat-form-field-empty mat-float mat-form-field-float"
                        for="mat-input-2"
                      >
                        Password 
                        
                        <span
                          aria-hidden="true"
                          class="mat-placeholder-required mat-form-field-required-marker ng-tns-c1-2"
                        >
                          *
                        </span>
                      </label>
                    </span>
                  </div>
                  
                </div>
                <div
                  class="mat-input-underline mat-form-field-underline"
                >
                  <span
                    class="mat-input-ripple mat-form-field-ripple"
                  />
                </div>
                <div
                  class="mat-input-subscript-wrapper mat-form-field-subscript-wrapper"
                  ng-reflect-ng-switch="hint"
                >
                  
                  
                  <div
                    class="mat-input-hint-wrapper mat-form-field-hint-wrapper ng-tns-c1-2 ng-trigger ng-trigger-transitionMessages"
                    style="opacity: 1;"
                  >
                    
                    <div
                      class="mat-input-hint-spacer mat-form-field-hint-spacer"
                    />
                  </div>
                </div>
              </div>
            </mat-form-field>
            
        
          </p>
          
  
        
          
                    
      
        
          <p
            class="registerButtons"
          >
            
          
            <button
              mat-button=""
              type="button"
            >
              Cancel
            </button>
             
          
            <button
              color="primary"
              mat-raised-button=""
              type="submit"
            >
              Create account
            </button>
              
        
          </p>
          
  
      
        </form>
        
    
      </mat-card-content>
      
  
    </mat-card>
    
  
  
  
  
    
  
  
  </hot-register-form>
</hot-register-page>
`;
